stats/opentelemetry/internal/tracing/custom_carrier.go

```diff
@@ -0,0 +1,141 @@
+/*
+ *
+ * Copyright 2024 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+// Package tracing implements the OpenTelemetry carrier for context propagation
+// in gRPC tracing.
+package tracing
+
+import (
+	"context"
+	"encoding/base64"
+	"strings"
+
+	"google.golang.org/grpc/grpclog"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/stats"
+)
+
+// GRPCTraceBinHeaderKey is the gRPC metadata header key `grpc-trace-bin` used
+// to propagate trace context in binary format.
+const GRPCTraceBinHeaderKey = "grpc-trace-bin"
+
+var logger = grpclog.Component("otel-plugin")
+
+// CustomCarrier is a TextMapCarrier that uses gRPC context to store and
+// retrieve any propagated key-value pairs in text format along with binary
+// format for `grpc-trace-bin` header.
+type CustomCarrier struct {
+	ctx context.Context
+}
+
+// NewCustomCarrier creates a new CustomMapCarrier with
+// the given context.
+func NewCustomCarrier(ctx context.Context) *CustomCarrier {
+	return &CustomCarrier{ctx: ctx}
+}
+
+// Get returns the string value associated with the passed key from the gRPC
+// context. It returns an empty string in any of the following cases:
+// - the key is not present in the context
+// - the value associated with the key is empty
+// - the key ends with "-bin" and is not `grpc-trace-bin`
+//
+// If the key is `grpc-trace-bin`, it retrieves the binary value using
+// `c.GetBinary()` and then base64 encodes it before returning. For all other
+// string keys, it retrieves the value from the context's metadata.
+func (c *CustomCarrier) Get(key string) string {
+	if key == GRPCTraceBinHeaderKey {
+		return base64.StdEncoding.EncodeToString(c.GetBinary())
```

Comment:
I asked this before, but why is this going through GetBinary? Just use stats package directly. This is an extra function call, I thinkkk the point of this conditional is to not go through the binary getters/setters optimization. In any case, it's extra complexity.

Reply:
Same reason as above. I have changed to stats package